pandas/core/indexes/datetimelike.py

```diff
@@ -527,24 +526,23 @@ def _maybe_box_as_values(self, values, **attribs):
         # - sort_values
         return values
 
+    @Appender(_index_shared_docs['astype'])
     def astype(self, dtype, copy=True):
-        if is_object_dtype(dtype):
-            return self._box_values_as_index()
-        elif is_string_dtype(dtype) and not is_categorical_dtype(dtype):
-            return Index(self.format(), name=self.name, dtype=object)
-        elif is_integer_dtype(dtype):
-            # TODO(DatetimeArray): use self._values here.
-            # Can't use ._values currently, because that returns a
-            # DatetimeIndex, which throws us in an infinite loop.
-            return Index(self.values.astype('i8', copy=copy), name=self.name,
-                         dtype='i8')
-        elif (is_datetime_or_timedelta_dtype(dtype) and
-              not is_dtype_equal(self.dtype, dtype)) or is_float_dtype(dtype):
-            # disallow conversion between datetime/timedelta,
-            # and conversions for any datetimelike to float
-            msg = 'Cannot cast {name} to dtype {dtype}'
-            raise TypeError(msg.format(name=type(self).__name__, dtype=dtype))
-        return super(DatetimeIndexOpsMixin, self).astype(dtype, copy=copy)
+        if is_dtype_equal(self.dtype, dtype) and copy is False:
+            # Ensure that self.astype(self.dtype) is self
+            return self
+
+        new_values = self._eadata.astype(dtype, copy=copy)
+
+        # we pass `dtype` to the Index constructor, for cases like
+        #  dtype=object to disable inference. But, DTA.astype ignores
+        #  integer sign and size, so we need to detect that case and
+        #  just choose int64.
+        dtype = pandas_dtype(dtype)
```

Comment:
not sure this is necessary as it already coerces properly, doing it here is very weird.

```
In [2]: pd.Index([1,2,3],dtype='int32')
Out[2]: Int64Index([1, 2, 3], dtype='int64')

```

Reply:
Not in the last 8 hours, no.  May need to wait on Tom to clarify, since all of this was taken from 24024.

(the fact that these things get closer attention in smaller doses reassures me that splitting is a good idea, even if it does cause rebasing hassles in the parent PR)

Reply:
yep, sounds ok then.

Reply:
What’s the question here? Why we do the integer check? Astype ignores the sign and size. I suppose the index constructo just ignores the size?

Reply:
the issue is I don't think you need the checking code / note as you have them here. This is *already* done in the Index constructor.

Reply:
index sign is ignored in `PeriodIndex.astype`, but not in the index constructor.

```python
In [6]: pd.period_range("2000", periods=4).astype("uint64")
Out[6]: Int64Index([10957, 10958, 10959, 10960], dtype='int64')

In [7]: pd.Index([0, 1], dtype='uint64')
Out[7]: UInt64Index([0, 1], dtype='uint64')
```

Reply:
Hmm never mind. Wrong branch.

```python
In [6]: pd.period_range("2000", periods=4).astype("uint64")
Out[6]: UInt64Index([10957, 10958, 10959, 10960], dtype='uint64')
```

In that case, this check can likely be omitted.

Reply:
Wait, now I'm confused. On 0.23.4 we have

```python
In [4]: pd.date_range('2000', periods=2).astype("u8")
Out[4]: Int64Index([946684800000000000, 946771200000000000], dtype='int64')

In [5]: pd.period_range('2000', periods=2).astype("u8")
Out[5]: Int64Index([10957, 10958], dtype='int64')

In [6]: pd.timedelta_range('2000', periods=2).astype("u8")
Out[6]: Int64Index([2000, 86400000002000], dtype='int64')
```

On master we have

```python
In [2]: pd.date_range('2000', periods=2).astype("u8")
Out[2]: Int64Index([946684800000000000, 946771200000000000], dtype='int64')

In [3]: pd.period_range('2000', periods=2).astype("u8")
Out[3]: UInt64Index([10957, 10958], dtype='uint64')

In [4]: pd.timedelta_range('2000', periods=2).astype("u8")
Out[4]: Int64Index([2000, 86400000002000], dtype='int64')
```

So period has changed from 0.23.4 (probably unintentionally). So we need to figure out which behavior we want and add tests / a release note for it.

Reply:
@TomAugspurger if you want to follow up on this ok, but this code should not be here, as its is properly in the constructor and *not* in astype.

Reply:
> but this code should not be here, as its is properly in the constructor and not in astype.
 
What is "proper" here? If we go by 0.23.4's behavior of always returning an Int64Index, then we have to have code here to detect that case. But if we want to return a UInt64Index then we can just pass dtype through to the Index constructor.

Reply:
having code here is just another inconsistency and maintenance burden. The constructor does the inference.

Reply:
OK I'm just going to decide that the expected behavior for {Datetime,Timedelta,Period}Index.astype("uint{8,16,32,64}")` is to return a UInt64Index.

@jbrockmendel do you want to do that here? It's not at all tested, and will need a release note.

Reply:
ok by me (of course its weird to do this, but hey)

Reply:
> do you want to do that here? It's not at all tested, and will need a release note.

I tried this, pretty much just deleting ten lines here, and ended up getting two failures in pandas/tests/indexes/interval/test_astype.py.  I can fix this by changing `dtype=dtype` to `dtype=new_values.dtype` in the call that wraps `self._eadata.astype`.  Is that what you have in mind?

Reply:
Attempt #2 at this also failed.  Any other ideas?

Reply:
Sorry I missed this note last night. I implemented this in 3fca8108a832fa7b80f317a75994ff8722b3743a if you could take a look.